### PR TITLE
[DO NOT MERGE] os/kernel: Add new variable to pthread_cond_s

### PR DESCRIPTION
**Pull request by rish-sg**
Description:

This patch fixes a race condition which exists between pthread_cond_wait() and pthread_cond_signal().
When a timer/signal interrupt occurs during the wait inside pthread_cond_wait, the signal causes sem_waitirq() to modify the semaphore count, leading to a lost signal.

----------------------------------------------------------------------

Root Cause:

In pthread_cond_signal(), the signal is only sent if sval < 0:

"Kernel/os/kernel/pthread/pthread_condsignal.c"
if (sval < 0) {
    ret = pthread_sem_give(&cond->sem);  // Signal sent
}
// If sval >= 0, signal is LOST!

The root cause of the issue is that `pthread_cond_signal()` checks the semaphore count (`semcount`) to determine if any thread is waiting. However, when a timer or signal interrupt occurs during the wait, `sem_waitirq()` increments the semaphore count, causing `pthread_cond_signal()` to incorrectly conclude that no thread is waiting and discard the signal.

----------------------------------------------------------------------

Logs:

sem_wait: sval after semcount--: -1, pid: 24, sem=0x60324b50    ← Thread blocked
timer_sigqueue: calling signo: 0 for pid: 24                    ← Interrupt
sem_waitirq: sval incremented to: 0, pid: 39, sem=0x60324b50    ← semcount: -1 → 0
pthread_cond_signal: sval=0, sem=0x60324b50                     ← sval = 0, NOT < 0!
pthread_cond_signal: Now sval is 0                              ← SIGNAL LOST!

----------------------------------------------------------------------

Proposed Fix:
Add an explicit waiter counter directly to the `pthread_cond_s` structure. This counter would be incremented before a thread starts waiting and decremented after it wakes up.
Since this counter is separate from the semaphore, it would not be modified by signal interrupts.

----------------------------------------------------------------------

### CODE REVIEW

**Comment by seokhun-eom24** [comment]
> Automated nightly Codex PR review
> PR: #7337 — [DO NOT MERGE] os/kernel: Add new variable to pthread_cond_s
> PR URL: https://github.com/Samsung/TizenRT/pull/7337
> GitHub updatedAt: 2026-05-13T08:45:39Z
> Refreshed (UTC): 2026-05-13 17:40:25 UTC
> Refreshed (KST): 2026-05-14 02:40:25 KST
> Comment model: single evolving Codex-authored PR comment
# PR #7337 — [DO NOT MERGE] os/kernel: Add new variable to pthread_cond_s

> **This review done by codex. AI reviews can be inaccurate.**  
> **UTC:** 2026-05-13 17:39 UTC  
> **KST (UTC+9):** 2026-05-14 02:39 KST

---

**Repository**: `Samsung/TizenRT`

**Base → Head**: `master` (`bf51b75c98fa67ba5762e86321b32a524398fc18`) → `pr/7337`

**HEAD Commit**: `c6b6bf624d4394c69b96bb7d48055c4c5d90f830`

**Scope**: pthread condition variable waiter tracking in `pthread_cond_wait()`, `pthread_cond_timedwait()`, `pthread_cond_signal()`, and `pthread_cond_broadcast()`.

---

## Review Summary

| Category | Status |
| --- | --- |
| Build / Compile Status | ⚠️ Not run |
| Functional Correctness | ❌ Blocking issue found |
| Validation Coverage | ❌ No regression test added for repeated signal/broadcast semantics |

**Final Verdict: ❗ Request Changes**

---

## Must-Fix Issues

<details>
<summary><strong>1. `waiters` can over-post the condition semaphore and wake a future waiter without a real signal</strong></summary>

<br>

| Item | Details |
| --- | --- |
| **Location** | `os/kernel/pthread/pthread_condsignal.c:117` |
| **Severity** | High |
| **Type** | Functional correctness |

**Problem**  
`pthread_cond_signal()` now uses `cond->waiters > 0` as the only guard before posting the condition semaphore. However, `cond->waiters` is decremented by the waiter only after `sem_wait()`/`pthread_sem_take()` returns. That leaves a window where a thread has already been signaled and unblocked, but `cond->waiters` is still positive. A second `pthread_cond_signal()` in that window can post the semaphore again even though there is no second blocked waiter.

**Impact**
- A stale positive semaphore count can remain on the condition variable.
- A later `pthread_cond_wait()` can consume that stale count and return immediately without a corresponding signal or broadcast.
- This regresses condition-variable semantics and can break existing users that rely on the mutex-protected predicate loop to sleep until a real state transition occurs.

**Evidence**
- `os/kernel/pthread/pthread_condsignal.c:117` — the new signal path posts whenever `cond->waiters > 0`.
- `os/kernel/pthread/pthread_condwait.c:137` — `cond->waiters` is decremented only after the waiter returns from `pthread_sem_take()`, so the count can still include a just-unblocked task.
- `os/kernel/semaphore/sem_post.c:217` — each post increments `sem->semcount`; an extra post can therefore leave a positive count for a future waiter.
- `os/kernel/semaphore/sem_wait.c:165` — a future waiter immediately succeeds when `semcount > 0`, consuming the stale signal instead of blocking.

**Required Fix**
Files to edit:
- `os/kernel/pthread/pthread_condsignal.c` — `pthread_cond_signal`
- `os/kernel/pthread/pthread_condbroadcast.c` — `pthread_cond_broadcast`
- `os/kernel/pthread/pthread_condwait.c` / `os/kernel/pthread/pthread_condtimedwait.c` — waiter accounting contract

Change outline:
- Do not use a raw waiter count as a reusable permission to post.
- Track waiters that are actually eligible to receive a signal, or consume the waiter accounting when the signal/broadcast posts.
- Keep the semaphore post count aligned with the number of currently blocked waiters so repeated signals cannot leave surplus tokens.
- Preserve the original race fix for `sem_waitirq()` by avoiding the old `semcount < 0` check as the sole wait-detection mechanism.

Example patch:
```diff
diff --git a/os/kernel/pthread/pthread_condsignal.c b/os/kernel/pthread/pthread_condsignal.c
@@
 	if (cond->waiters > 0) {
+		cond->waiters--;
 		ret = pthread_sem_give((sem_t *)&cond->sem);
 	}
diff --git a/os/kernel/pthread/pthread_condwait.c b/os/kernel/pthread/pthread_condwait.c
@@
-		/* Decrement waiters counter */
-		flags = enter_critical_section();
-		cond->waiters--;
-		leave_critical_section(flags);
+		/* Only decrement here if the wait ended without a signal consuming
+		 * the waiter slot, for example cancellation or error recovery.
+		 */
```

Inference:
- The exact final shape needs a small state distinction between "signal consumed this waiter slot" and "wait ended by timeout/cancellation/error." The example is intentionally partial because blindly moving the decrement would break timeout and cancellation accounting.

**Validation Method**
- Add a regression test with one waiting thread and two immediate `pthread_cond_signal()` calls before the waiter can run again; then start a second waiter and verify it blocks until a new signal.
- Add the same coverage for `pthread_cond_broadcast()` followed by another signal/broadcast cycle to prove no surplus semaphore count remains.

</details>

---

## Nice-to-Have Improvements

<details>
<summary><strong>1. Add targeted regression coverage for the reported lost-signal race and the new over-post boundary</strong></summary>

<br>

| Item | Details |
| --- | --- |
| **Location** | `apps/examples/testcase/le_tc/kernel/tc_pthread.c` |
| **Severity** | Medium |
| **Type** | Validation |

**Problem**  
The PR changes a core pthread synchronization primitive but does not add a regression test for either the reported `sem_waitirq()` lost-signal case or the new waiter-count edge cases.

**Impact**
- The intended fix could regress later because the failure requires precise timing around signal/timeout interruption.
- Repeated signal and broadcast semantics are not protected by automated coverage.

**Recommended Action**
- Add a pthread testcase that forces the waiter into `pthread_cond_wait()`/`pthread_cond_timedwait()`, injects the signal/timer interruption path, and confirms the subsequent `pthread_cond_signal()` is not lost.
- Add a second testcase that calls `pthread_cond_signal()` twice for one waiter and verifies a later waiter does not consume a stale semaphore token.

**Example Validation Matrix**

| Area | Example Target / Check | Result |
| --- | --- | --- |
| Lost-signal race | Forced signal/timer interruption while waiting | Not Run |
| Repeated signal | One waiter, two signals, then a second waiter | Not Run |
| Broadcast boundary | N waiters, broadcast, then new waiter blocks | Not Run |

</details>

---

<details>
<summary><strong>Notable Improvements</strong></summary>

<br>

### ✔ The PR targets the real weak point in the old implementation

- The old `pthread_cond_signal()` path depended on the condition semaphore count, and `sem_waitirq()` explicitly adjusts that count when a signal or timeout interrupts a wait. Introducing condition-variable-owned waiter state is a reasonable direction for separating "a thread intends to wait" from the semaphore's transient count.

### ✔ Static and dynamic initializers were both updated

- `pthread_cond_init()` initializes `cond->waiters`, and `PTHREAD_COND_INITIALIZER` was expanded to include the new field.

</details>

---

## Final Assessment

### Must-Fix Summary

- The current waiter counter can authorize more semaphore posts than there are blocked waiters, leaving a stale condition signal for a future wait.

### Nice-to-Have Summary

- Add regression tests for both the original lost-signal race and repeated signal/broadcast edge cases.

### Residual Risk / Test Gap

- This was a static review only; no TizenRT build or runtime pthread testcase was run in this unattended pass.
- The suggested fix shape is partial and should be validated against timeout, cancellation, and broadcast interactions before merge.

---

## Final Verdict

> ❗ Request Changes

The PR addresses a plausible race, but the new raw waiter counter introduces an over-post path that can wake future waiters without a corresponding signal. Please fix the accounting contract and add targeted synchronization regression tests before merging.

**Review by ewoodev — DISMISSED** [review]

File: os/kernel/pthread/pthread_condtimedwait.c [hunk 1]
```diff
@@ -288,15 +288,21 @@ int pthread_cond_timedwait(FAR pthread_cond_t *cond, FAR pthread_mutex_t *mutex,
 
 						wd_start(rtcb->waitdog, ticks, (wdentry_t)pthread_condtimedout, 2, (uint32_t)mypid, (uint32_t)SIGCONDTIMEDOUT);
 
-						/* Take the condition semaphore.  Do not restore interrupts
-						 * until we return from the wait.  This is necessary to
-						 * make sure that the watchdog timer and the condition wait
-						 * are started atomically.
-						 */
+					/* Increment the waiter count */
```
Comment (seokhun-eom24) [inline os/kernel/pthread/pthread_condtimedwait.c]: Could you correct the indentation?

File: os/kernel/pthread/pthread_condwait.c [hunk 2]
```diff
@@ -128,6 +134,11 @@ int pthread_cond_wait(FAR pthread_cond_t *cond, FAR pthread_mutex_t *mutex)
 
 		sched_unlock();
 
+		/* Decrement waiters counter */
```
Comment (seokhun-eom24) [inline os/kernel/pthread/pthread_condwait.c]: I think in this point, other thread can send signal again, and call `pthread_sem_give` again.
Because cond->waiters is not decreased yet.
How about decrease cond->wainters in pthread_condsignal.c and pthread_condbroadcast.c?

**Review by seokhun-eom24 — COMMENTED** [review]
Could you rebase latest master branch and test again?
Recently we updated some semaphore related codes.

**Review by ewoodev — APPROVED** [review]

**Review by seokhun-eom24 — APPROVED** [review]
LGTM

**Review by ewoodev — APPROVED** [review]

**Review by allen-kim-sec — APPROVED** [review]
looks good to me... it's nice

**Comment by vivek1-j** [comment]
LGTM
